File: helpers/fenwick.py
#!/usr/bin/env python3


class FenwickTree(object):
    def __init__(self, nums):
        self.nums = [0] * (len(nums) + 1)
        for i in range(1, len(self.nums)):
            self.add(i, nums[i-1])

    def add(self, i, n):
        while i < len(self.nums):
            self.nums[i] += n
            i += (i & -i)
    
    def sum(self, i):
        result = 0
        while i:
            result += self.nums[i]
            i -= (i & -i)
        return result

File: helpers/datastructures/cpp/fenwick.cpp
#include <vector>

using namespace std;

struct FenwickTree {
    vector<int> v;

    FenwickTree(vector<int> u)
    {
        v = vector<int>(u.size() + 1);
        for (int i = 0; i < u.size(); i++)
            add(i+1, u[i]);
    }

    void add(int i, int n)
    {
        while (i < v.size()) {
            v[i] += n;
            i += (i & -i);
        }
    }

    int sum(int i)
    {
        int result = 0;
        while (i) {
            result += v[i];
            i -= (i & -i);
        }
        return result;
    }
};
